### PR TITLE
[DIS] DIS-90 Add declarative transactions to DIS service

### PR DESCRIPTION
**Pull request by jimmordino**
Adds a fully functional `@Transactional` annotation that specifies, among other things, isolation level and read-only status.  Also adds a `@SiteIdSource` annotation that decorates one and only argument of a transactional method to tells the method interceptor how to obtain the site it will use to determine which database shard to connect with.

- [x] Add `@Transaction` and `@SiteIdSource` annotations to all 'front door' service methods
- [x] Update transactional method interceptor to use annotations to determine which db (shard, master vs. slave) will service the request
- [x] Update DaoFactory singleton to store the shard name and read-only status in a `ThreadLocal` (as MyBatis does with the session/connection)
- [x] Add new Exceptions for invalid configuration and missing required annotation
- [x] Add unit tests for all new functionality
- [x] Update build configuration to use a freshly cloned DB for `integrationTest` target

### CODE REVIEW

File: service/discussion/src/main/java/com/wikia/discussionservice/dao/mybatis/MyBatisDaoFactory.java [hunk 1]
```diff
@@ -20,10 +26,59 @@ public MyBatisDaoFactory(@Named("sessionManagerMap")
...
+    String currentShardName = shardMap.get(SHARD_NAME_KEY);
+    if (!checkShardName.equals(currentShardName)) {
+      throw new IllegalStateException(String.format(
+          "Current shard name %s does not match expected shard name %s",
+          currentShardName, checkShardName));
+    }
+    // Ensure that we're not attempting a read-write operation if the session was
+    // created read-only
+    boolean currentReadOnly = Boolean.valueOf(shardMap.get(READONLY_KEY));
+    if (currentReadOnly && !checkReadOnly) {
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/dao/mybatis/MyBatisDaoFactory.java]: Looking at the variable name, first guess is that this is a flag for whether or not readOnly check should be enforced, which probably is an incorrect guess. Maybe a comment might help me or another name :)

File: service/discussion/src/main/java/com/wikia/discussionservice/exceptions/InvalidConfigurationException.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package com.wikia.discussionservice.exceptions;
+
+public class InvalidConfigurationException extends RuntimeException {
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/exceptions/InvalidConfigurationException.java]: Should this be a `DiscussionServiceException`?

File: service/discussion/src/main/java/com/wikia/discussionservice/exceptions/MissingAnnotationException.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.wikia.discussionservice.exceptions;
+
+public class MissingAnnotationException extends RuntimeException {
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/exceptions/MissingAnnotationException.java]: Should this be a `DiscussionServiceException`?

File: service/discussion/src/main/java/com/wikia/discussionservice/resources/VoteResource.java [hunk 4]
```diff
@@ -82,7 +82,7 @@ public VoteResource(VoteService voteService, VoteRepresentationMapper voteMapper
       @ApiResponse(code = 400, message = "Bad request", response = HalProblem.class),
       @ApiResponse(code = 401, message = "Unauthorized", response = HalProblem.class),
       @ApiResponse(code = 403, message = "Forbidden", response = HalProblem.class),
-      @ApiResponse(code = 404, message = "Post not found", response = HalProblem.class)})
+      @ApiResponse(code = 404, message = "Post not foundÏÏ", response = HalProblem.class)})
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/resources/VoteResource.java]: What are these characters? :)

File: service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java [hunk 5]
```diff
@@ -43,7 +44,13 @@ public ForumService(DaoFactory daoFactory) {
     this.daoFactory = daoFactory;
   }
 
-  public Optional<Forum> getRootForum(Integer siteId) {
+  @Transactional(
+      executorType = ExecutorType.SIMPLE,
+      isolation = Isolation.READ_COMMITTED,
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java]: Should this be `READ_COMMITTED` or `SERIALIZABLE`?

File: service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java [hunk 6]
```diff
@@ -80,73 +87,102 @@ public Forum convertForumListToRootForum(List<Forum> forumList) {
...
+  public Optional<Forum> getForum(User user, @SiteIdSource Integer siteId, Long forumId) {
     return getForum(user, siteId, forumId, SHALLOW_FORUM_LIMIT_DEFAULT, 0);
   }
 
-  public Optional<Forum> getForum(User user, Integer siteId, Long forumId,
+  @Transactional(
+      executorType = ExecutorType.SIMPLE,
+      isolation = Isolation.SERIALIZABLE,
+      rethrowExceptionsAs = DiscussionServiceException.class,
+      exceptionMessage = "Unable to get forum for site ID %d with ID %d",
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java]: Should the placeholder orders be the same as param list order?

File: service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java [hunk 7]
```diff
@@ -80,73 +87,102 @@ public Forum convertForumListToRootForum(List<Forum> forumList) {
...
   @Transactional(
       executorType = ExecutorType.SIMPLE,
-      isolation = Isolation.SERIALIZABLE,
+      isolation = Isolation.READ_COMMITTED,
       rethrowExceptionsAs = DiscussionServiceException.class,
-      exceptionMessage = "Unable to create forum with site ID %s"
-      )
-  public Optional<Forum> createForum(User user, Forum forum) {
+      exceptionMessage = "Unable to create forum with site ID %s")
+  public Optional<Forum> createForum(User user, @SiteIdSource Forum forum) {
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java]: missing siteId param

File: service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java [hunk 8]
```diff
@@ -80,73 +87,102 @@ public Forum convertForumListToRootForum(List<Forum> forumList) {
...
   @Transactional(
       executorType = ExecutorType.SIMPLE,
-      isolation = Isolation.SERIALIZABLE,
+      isolation = Isolation.READ_COMMITTED,
       rethrowExceptionsAs = DiscussionServiceException.class,
-      exceptionMessage = "Unable to update forum with site ID %s"
-      )
-  public Optional<Forum> updateForum(User user, Forum forum) {
+      exceptionMessage = "Unable to update forum with site ID %s")
+  public Optional<Forum> updateForum(User user, @SiteIdSource  Forum forum) {
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java]: missing siteId param

File: service/discussion/src/main/java/com/wikia/discussionservice/services/PostService.java [hunk 9]
```diff
@@ -58,16 +64,13 @@ public PostService(DaoFactory daoFactory) {
...
-   */
+   **/
   @Transactional(
       executorType = ExecutorType.SIMPLE,
       isolation = Isolation.READ_COMMITTED,
       rethrowExceptionsAs = DiscussionServiceException.class,
       exceptionMessage = "Unable to create post")
-  public Optional<Post> createPost(User user, Post post)
-      throws DiscussionServiceException {
+  public Optional<Post> createPost(User user, @SiteIdSource Post post) {
```
Comment (armonr) [inline service/discussion/src/main/java/com/wikia/discussionservice/services/PostService.java]: missing siteId param

File: service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java (hunk 6 above)
Comment (jimmordino) [inline service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java]: Yes, user was added later - good catch

File: service/discussion/src/main/java/com/wikia/discussionservice/dao/mybatis/MyBatisDaoFactory.java (hunk 1 above)
Comment (jimmordino) [inline service/discussion/src/main/java/com/wikia/discussionservice/dao/mybatis/MyBatisDaoFactory.java]: Fair point, hope this looks better 9c4297133eaa1368a189e8dec42e0b9535dbca8c

File: service/discussion/src/main/java/com/wikia/discussionservice/exceptions/InvalidConfigurationException.java (hunk 2 above)
Comment (jimmordino) [inline service/discussion/src/main/java/com/wikia/discussionservice/exceptions/InvalidConfigurationException.java]: Sorry, brain fade.  0a8300aa328fe5eb1fe03048bae462f1e07cb7fd

File: service/discussion/src/main/java/com/wikia/discussionservice/exceptions/MissingAnnotationException.java (hunk 3 above)
Comment (jimmordino) [inline service/discussion/src/main/java/com/wikia/discussionservice/exceptions/MissingAnnotationException.java]: Sorry, brain fade.  0a8300aa328fe5eb1fe03048bae462f1e07cb7fd

File: service/discussion/src/main/java/com/wikia/discussionservice/resources/VoteResource.java (hunk 4 above)
Comment (jimmordino) [inline service/discussion/src/main/java/com/wikia/discussionservice/resources/VoteResource.java]: probably from typing in the wrong window - thanks for catching this.  5c0fb96f8a2836ba848a32d7cd00048e8a5cd96f

File: service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java (hunk 5 above)
Comment (jimmordino) [inline service/discussion/src/main/java/com/wikia/discussionservice/services/ForumService.java]: Good catch: 1f4c3e542d0eaaa87c3fcd566a03a2161c3c444c

**Comment by armonr** [comment]
:+1: :+1: 
:shipit:

**Pull request merged**